File: Graphs/graph-kahn.java
// Topological Sort (Kahn's Algorithm)
// Type: DAG, BFS-based
// Time: O(V + E)
// Space: O(V + E)

import java.util.*;

public class KahnTopoSort {
    public static List<Integer> topologicalSort(Map<Integer, List<Integer>> graph) {
        Map<Integer, Integer> indegree = new HashMap<>();
        for (int u : graph.keySet())
            for (int v : graph.get(u))
                indegree.put(v, indegree.getOrDefault(v, 0) + 1);

        Queue<Integer> queue = new LinkedList<>();
        for (int u : graph.keySet())
            if (!indegree.containsKey(u))
                queue.offer(u);

        List<Integer> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            int u = queue.poll();
            result.add(u);
            for (int v : graph.getOrDefault(u, new ArrayList<>())) {
                indegree.put(v, indegree.get(v) - 1);
                if (indegree.get(v) == 0)
                    queue.offer(v);
            }
        }

        if (result.size() != graph.size())
            throw new RuntimeException("Cycle detected");

        return result;
    }
}

File: Graphs/graph-kahn.c
// Topological Sort (Kahn's Algorithm)
// Type: DAG, BFS-based
// Time: O(V + E)
// Space: O(V + E)

#include <stdio.h>
#include <stdlib.h>

#define MAX 100

int graph[MAX][MAX], indegree[MAX], V;

void kahnTopologicalSort() {
    int queue[MAX], front = 0, rear = 0, count = 0;
    int result[MAX];

    for (int i = 0; i < V; i++)
        if (indegree[i] == 0)
            queue[rear++] = i;

    while (front < rear) {
        int u = queue[front++];
        result[count++] = u;

        for (int v = 0; v < V; v++) {
            if (graph[u][v]) {
                indegree[v]--;
                if (indegree[v] == 0)
                    queue[rear++] = v;
            }
        }
    }

    if (count != V)
        printf("Cycle detected\n");
    else {
        for (int i = 0; i < V; i++)
            printf("%d ", result[i]);
        printf("\n");
    }
}
